File: baseten-performance-client/node_bindings/__test__/test_methods.js
// Concise parameterized test for HTTP method functionality
const { PerformanceClient } = require('./index.js');

// Test configuration
const testCases = [
  { method: 'GET', path: '/get' },
  { method: 'POST', path: '/post' },
  { method: 'PUT', path: '/put' },
  { method: 'PATCH', path: '/patch' },
  { method: 'DELETE', path: '/delete' },
  { method: null, path: '/post' }, // Default to POST
];

async function runTests() {
  console.log('🧪 Starting HTTP method tests...\n');

  const client = new PerformanceClient('https://httpbin.org', 'test-key');
  const payloads = [{ test: 'data' }];

  let passed = 0;
  let failed = 0;

  // Test valid methods
  for (const { method, path } of testCases) {
    try {
      const args = method ? [path, payloads, undefined, undefined, method] : [path, payloads];
      const response = await client.batchPost(...args);

      if (response.data.length === 1 && response.total_time > 0) {
        console.log(`✅ ${method || 'default'} -> ${path}`);
        passed++;
      } else {
        console.log(`❌ ${method || 'default'} -> Invalid response`);
        failed++;
      }
    } catch (e) {
      console.log(`❌ ${method || 'default'} -> ${e.message}`);
      failed++;
    }
  }

  // Test invalid method
  try {
    await client.batchPost('/post', payloads, undefined, undefined, 'INVALID');
    console.log('❌ INVALID method -> Should have failed');
    failed++;
  } catch (e) {
    if (e.message.includes('Invalid HTTP method')) {
      console.log('✅ INVALID method -> Correctly rejected');
      passed++;
    } else {
      console.log(`❌ INVALID method -> Wrong error: ${e.message}`);
      failed++;
    }
  }

  console.log(`\n📊 Results: ${passed} passed, ${failed} failed`);

  if (failed === 0) {
    console.log('🎉 All tests passed!');
    process.exit(0);
  } else {
    console.log('💥 Some tests failed!');
    process.exit(1);
  }
}

runTests().catch(console.error);

File: baseten-performance-client/python_bindings/tests/test_methods.py
#!/usr/bin/env python3
"""
Concise parameterized test for HTTP method functionality
"""

import pytest
from baseten_performance_client import PerformanceClient


@pytest.mark.parametrize(
    "method,expected_path",
    [
        ("GET", "/get"),
        ("POST", "/post"),
        ("PUT", "/put"),
        ("PATCH", "/patch"),
        ("DELETE", "/delete"),
        ("HEAD", "/get"),  # HEAD to /get returns headers only
        ("OPTIONS", "/"),  # OPTIONS to root returns CORS headers
        (None, "/post"),  # Default to POST
    ],
)
def test_http_methods(method, expected_path):
    """Test all HTTP methods including default behavior"""
    client = PerformanceClient(base_url="https://httpbin.org", api_key="test-key")
    payloads = [{"test": "data"}]

    kwargs = {"method": method} if method else {}
    response = client.batch_post(expected_path, payloads, **kwargs)

    assert len(response.data) == 1
    assert response.total_time > 0


def test_invalid_method():
    """Test that invalid methods are rejected"""
    client = PerformanceClient(base_url="https://httpbin.org", api_key="test-key")

    with pytest.raises(ValueError, match="Invalid HTTP method"):
        client.batch_post("/post", [{"test": "data"}], method="INVALID")


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
